pcsx2/DEV9/ATA/ATA_Transfer.cpp

```diff
@@ -98,18 +107,98 @@ bool ATA::IO_Write()
 	}
 
 	hddImage.seekp(entry.sector * 512, std::ios::beg);
-	hddImage.write((char*)entry.data, entry.length);
-	if (hddImage.fail())
+	if (hddSparse)
 	{
-		Console.Error("DEV9: ATA: File write error");
-		pxAssert(false);
-		abort();
+		//What size block is best (must fit evenly into 512 bytes)?
+		const int sparseSize = 512;
+		const static u8 zeroBlock[sparseSize]{0};
+		for (size_t i = 0; i < entry.length; i += sparseSize)
+		{
+			if (memcmp(&entry.data[i], zeroBlock, sparseSize) == 0)
```

Comment:
I feel like there has to be a better way to do this than zero a block of stack and memcmp, especially for non-fixed block size

stackoverflow says `!x[0] && memcmp(x, x+1, len-1) == 0`
or you can have some fun with sse
```cpp
bool all_zero_aligned64(const void* data, size_t len)
{
	using V = __m128i;
	constexpr int cacheline = 64 / sizeof(V);
	const V* src = static_cast<const V*>(data);
	const V* end = reinterpret_cast<const V*>(static_cast<const char*>(data) + len);
	for (; src < end; src += cacheline)
	{
		V acc = _mm_setzero_si128();
		for (int j = 0; j < cacheline; j++)
			acc = _mm_or_si128(acc, _mm_load_si128(src + j));
		if (65535 == _mm_movemask_epi8(_mm_cmpeq_epi8(acc, _mm_setzero_si128())))
			return false;
	}
	return true;
}
```

Reply:
> stackoverflow says `!x[0] && memcmp(x, x+1, len-1) == 0`

I did do a search before writing that code to see if there was a better way. 
In one of those answers they benchmarked that method against memcmp on a temporary stack block (like I'm using) and found it to be 4x slower.

> or you can have some fun with sse

I wasn't sure about including a large amount of code to perform this task, but it seems like your suggesting that is the way to go.

Reply:
I couldn't get the sse version to work correctly, It was giving me false positives for some reason
I grabbed another implementation from stackoverflow.